File: test/ch21-benchmarks/bench-004-binary-tree.go
// BENCH-004: Binary tree allocation/deallocation - Go
// Tests: memory allocator, GC, pointer chasing
package main

type TreeNode struct {
	left  *TreeNode
	right *TreeNode
}

func makeTree(depth int) *TreeNode {
	if depth <= 0 {
		return &TreeNode{}
	}
	return &TreeNode{
		left:  makeTree(depth - 1),
		right: makeTree(depth - 1),
	}
}

func checkTree(node *TreeNode) int {
	if node.left == nil {
		return 1
	}
	return 1 + checkTree(node.left) + checkTree(node.right)
}

func main() {
	maxDepth := 16
	minDepth := 4

	// Stretch tree
	stretchDepth := maxDepth + 1
	stretchTree := makeTree(stretchDepth)
	stretchCheck := checkTree(stretchTree)

	// Long-lived tree
	longLivedTree := makeTree(maxDepth)

	// Create and destroy many trees
	totalChecks := 0
	for depth := minDepth; depth <= maxDepth; depth += 2 {
		iterations := 1 << uint(maxDepth-depth+minDepth)
		for i := 0; i < iterations; i++ {
			tree := makeTree(depth)
			totalChecks += checkTree(tree)
		}
	}

	// Final checksum
	longCheck := checkTree(longLivedTree)

	_ = stretchCheck
	_ = totalChecks
	_ = longCheck
	// Silent for benchmarking
}

File: test/ch21-benchmarks/bench-004-binary-tree.ts
// BENCH-004: Binary tree allocation/deallocation - Deno TypeScript
// Tests: memory allocator, V8 GC, pointer chasing

class TreeNode {
    left: TreeNode | null;
    right: TreeNode | null;

    constructor(left: TreeNode | null = null, right: TreeNode | null = null) {
        this.left = left;
        this.right = right;
    }
}

function makeTree(depth: number): TreeNode {
    if (depth <= 0) {
        return new TreeNode();
    }
    return new TreeNode(makeTree(depth - 1), makeTree(depth - 1));
}

function checkTree(node: TreeNode): number {
    if (node.left === null) {
        return 1;
    }
    return 1 + checkTree(node.left) + checkTree(node.right!);
}

function main(): void {
    const maxDepth = 16;
    const minDepth = 4;

    // Stretch tree
    const stretchDepth = maxDepth + 1;
    const stretchTree = makeTree(stretchDepth);
    const stretchCheck = checkTree(stretchTree);

    // Long-lived tree
    const longLivedTree = makeTree(maxDepth);

    // Create and destroy many trees
    let totalChecks = 0;
    for (let depth = minDepth; depth <= maxDepth; depth += 2) {
        const iterations = 1 << (maxDepth - depth + minDepth);
        for (let i = 0; i < iterations; i++) {
            const tree = makeTree(depth);
            totalChecks += checkTree(tree);
        }
    }

    // Final checksum
    const longCheck = checkTree(longLivedTree);

    // Silent for benchmarking
}

main();
